File: api/services/processor.js
const { spawn, execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const { CONTAINER_MIMES, AUDIO_MIMES } = require('../config/constants');
const { cleanupJobFiles, sanitizeFilename } = require('../utils/files');
const {
  activeDownloads,
  activeProcesses,
  activeJobsByType,
  removePendingJob,
  unlinkJobFromClient,
  sendProgress
} = require('./state');

async function processVideo(inputPath, outputPath, opts = {}) {
  const {
    isAudio = false,
    isGif = false,
    audioFormat = 'mp3',
    audioBitrate = '320',
    container = 'mp4',
    jobId = ''
  } = opts;

  const inputExt = path.extname(inputPath).slice(1).toLowerCase();
  const outputExt = isGif ? 'gif' : (isAudio ? audioFormat : container);

  if (!isAudio && !isGif && inputExt === outputExt) {
    console.log(`[${jobId}] Format match (${inputExt}), skipping ffmpeg`);
    return { path: inputPath, ext: inputExt, skipped: true };
  }

  const ffmpegArgs = ['-y', '-i', inputPath];

  if (isAudio) {
    if (audioFormat === 'mp3') {
      ffmpegArgs.push('-codec:a', 'libmp3lame', '-b:a', `${audioBitrate}k`);
    } else if (audioFormat === 'm4a') {
      ffmpegArgs.push('-codec:a', 'aac', '-b:a', `${audioBitrate}k`);
    } else if (audioFormat === 'opus') {
      ffmpegArgs.push('-codec:a', 'libopus', '-b:a', `${audioBitrate}k`);
    } else if (audioFormat === 'wav') {
      ffmpegArgs.push('-codec:a', 'pcm_s16le');
    } else if (audioFormat === 'flac') {
      ffmpegArgs.push('-codec:a', 'flac');
    } else {
      ffmpegArgs.push('-codec:a', 'copy');
    }
  } else if (isGif) {
    ffmpegArgs.length = 0;
    ffmpegArgs.push(
      '-y',
      '-i', inputPath,
      '-vf', 'fps=15,scale=480:-1:flags=lanczos,split[s0][s1];[s0]palettegen[p];[s1][p]paletteuse',
      '-loop', '0'
    );
  } else {
    ffmpegArgs.push('-codec', 'copy');
    if (container === 'mp4' || container === 'mov') {
      ffmpegArgs.push('-movflags', '+faststart');
    }
  }

  ffmpegArgs.push(outputPath);

  await new Promise((resolve, reject) => {
    const ffmpeg = spawn('ffmpeg', ffmpegArgs);
    let ffmpegStderr = '';

    ffmpeg.stderr.on('data', (data) => {
      const msg = data.toString();
      ffmpegStderr += msg;
      if (msg.includes('Error') || msg.includes('error')) {
        console.error(`[${jobId}] ffmpeg: ${msg.trim()}`);
      }
    });

    ffmpeg.on('close', (code) => {
      if (code === 0) resolve();
      else {
        console.error(`[${jobId}] FFmpeg failed with code ${code}. Last 500 chars: ${ffmpegStderr.slice(-500)}`);
        reject(new Error(`Encoding failed (code ${code})`));
      }
    });

    ffmpeg.on('error', (err) => {
      console.error(`[${jobId}] FFmpeg spawn error: ${err.message}`);
      reject(err);
    });
  });

  return { path: outputPath, ext: outputExt, skipped: false };
}

async function streamFile(res, req, filePath, opts = {}) {
  const {
    filename = 'download',
    ext = 'mp4',
    mimeType = 'video/mp4',
    downloadId,
    url = '',
    jobType = 'download',
    onCleanup = null
  } = opts;

  const stat = fs.statSync(filePath);
  const safeFilename = sanitizeFilename(filename);
  const fullFilename = `${safeFilename}.${ext}`;
  const asciiFilename = safeFilename.replace(/[^\x20-\x7E]/g, '_') + '.' + ext;

  sendProgress(downloadId, 'sending', 'Sending file to you...');

  res.setHeader('Content-Type', mimeType);
  res.setHeader('Content-Length', stat.size);
  res.setHeader('Content-Disposition', `attachment; filename="${asciiFilename}"; filename*=UTF-8''${encodeURIComponent(fullFilename)}`);

  const stream = fs.createReadStream(filePath);
  let finished = false;

  stream.pipe(res);

  const cleanup = () => {
    if (finished) return;
    finished = true;
    sendProgress(downloadId, 'complete', 'Download complete!');
    activeDownloads.delete(downloadId);
    activeProcesses.delete(downloadId);
    removePendingJob(downloadId);
    activeJobsByType[jobType]--;
    unlinkJobFromClient(downloadId);

    console.log(`[Queue] ${jobType} finished. Active: ${JSON.stringify(activeJobsByType)}`);
    if (onCleanup) onCleanup();
    setTimeout(() => cleanupJobFiles(downloadId), 2000);
  };

  stream.on('close', cleanup);
  res.on('finish', cleanup);

  stream.on('error', (err) => {
    if (finished) return;
    finished = true;
    console.error(`[${downloadId}] Stream error:`, err);
    sendProgress(downloadId, 'error', 'Failed to send file');
    activeProcesses.delete(downloadId);
    removePendingJob(downloadId);
    activeJobsByType[jobType]--;
    unlinkJobFromClient(downloadId);
    setTimeout(() => cleanupJobFiles(downloadId), 2000);
  });

  req.on('close', () => {
    if (finished) return;
    finished = true;
    stream.destroy();
    activeProcesses.delete(downloadId);
    removePendingJob(downloadId);
    activeJobsByType[jobType]--;
    unlinkJobFromClient(downloadId);
    console.log(`[Queue] ${jobType} cancelled. Active: ${JSON.stringify(activeJobsByType)}`);
    setTimeout(() => {
      try { fs.unlinkSync(filePath); } catch {}
    }, 1000);
  });
}

function probeForGif(filePath) {
  try {
    const probeResult = execSync(
      `ffprobe -v quiet -print_format json -show_streams -show_format "${filePath}"`,
      { encoding: 'utf8' }
    );
    const probe = JSON.parse(probeResult);
    const hasAudio = probe.streams?.some(s => s.codec_type === 'audio');
    const duration = parseFloat(probe.format?.duration || '999');
    return !hasAudio && duration < 60;
  } catch (e) {
    return false;
  }
}

function getMimeType(ext, isAudio, isGif) {
  if (isGif) return 'image/gif';
  if (isAudio) return AUDIO_MIMES[ext] || 'audio/mpeg';
  return CONTAINER_MIMES[ext] || 'video/mp4';
}

module.exports = {
  processVideo,
  streamFile,
  probeForGif,
  getMimeType
};

File: internal/services/processor.go
package services

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/coah80/yoink/internal/config"
	"github.com/coah80/yoink/internal/util"
)

type ProcessVideoOpts struct {
	IsAudio      bool
	IsGif        bool
	AudioFormat  string
	AudioBitrate string
	Container    string
	JobID        string
}

type ProcessResult struct {
	Path    string
	Ext     string
	Skipped bool
}

func ProcessVideo(inputPath, outputPath string, opts ProcessVideoOpts) (*ProcessResult, error) {
	inputExt := strings.ToLower(strings.TrimPrefix(filepath.Ext(inputPath), "."))
	outputExt := opts.Container
	if opts.IsGif {
		outputExt = "gif"
	} else if opts.IsAudio {
		outputExt = opts.AudioFormat
	}

	if !opts.IsAudio && !opts.IsGif && inputExt == outputExt {
		log.Printf("[%s] Format match (%s), skipping ffmpeg", opts.JobID, inputExt)
		return &ProcessResult{Path: inputPath, Ext: inputExt, Skipped: true}, nil
	}

	args := []string{"-y", "-i", inputPath}

	if opts.IsAudio {
		bitrate := opts.AudioBitrate
		if bitrate == "" {
			bitrate = "320"
		}
		switch opts.AudioFormat {
		case "mp3":
			args = append(args, "-codec:a", "libmp3lame", "-b:a", bitrate+"k")
		case "m4a":
			args = append(args, "-codec:a", "aac", "-b:a", bitrate+"k")
		case "opus":
			args = append(args, "-codec:a", "libopus", "-b:a", bitrate+"k")
		case "wav":
			args = append(args, "-codec:a", "pcm_s16le")
		case "flac":
			args = append(args, "-codec:a", "flac")
		default:
			args = append(args, "-codec:a", "copy")
		}
	} else if opts.IsGif {
		args = []string{"-y", "-i", inputPath,
			"-vf", "fps=15,scale=480:-1:flags=lanczos,split[s0][s1];[s0]palettegen[p];[s1][p]paletteuse",
			"-loop", "0",
		}
	} else {
		args = append(args, "-codec", "copy")
		if opts.Container == "mp4" || opts.Container == "mov" {
			args = append(args, "-movflags", "+faststart")
		}
	}

	args = append(args, outputPath)

	cmd := exec.Command("ffmpeg", args...)
	stderrPipe, _ := cmd.StderrPipe()

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("failed to start ffmpeg: %w", err)
	}

	stderrBytes, _ := io.ReadAll(stderrPipe)
	if err := cmd.Wait(); err != nil {
		errStr := string(stderrBytes)
		if len(errStr) > 500 {
			errStr = errStr[len(errStr)-500:]
		}
		log.Printf("[%s] FFmpeg failed. Last 500 chars: %s", opts.JobID, errStr)
		return nil, fmt.Errorf("Encoding failed (code %d)", cmd.ProcessState.ExitCode())
	}

	return &ProcessResult{Path: outputPath, Ext: outputExt, Skipped: false}, nil
}

func StreamFile(w http.ResponseWriter, r *http.Request, filePath string, filename, ext, mimeType, downloadID, sourceURL, jobType string, onCleanup func()) {
	info, err := os.Stat(filePath)
	if err != nil {
		http.Error(w, "File not found", 404)
		return
	}

	safeFilename := util.SanitizeFilename(filename)
	fullFilename := safeFilename + "." + ext
	asciiFilename := toASCII(safeFilename) + "." + ext

	Global.SendProgressSimple(downloadID, "sending", "Sending file to you...")

	w.Header().Set("Content-Type", mimeType)
	w.Header().Set("Content-Length", fmt.Sprintf("%d", info.Size()))
	w.Header().Set("Content-Disposition",
		fmt.Sprintf(`attachment; filename="%s"; filename*=UTF-8''%s`,
			asciiFilename, url.PathEscape(fullFilename)))

	f, err := os.Open(filePath)
	if err != nil {
		log.Printf("[%s] Failed to open file for streaming: %v", downloadID, err)
		Global.SendProgressSimple(downloadID, "error", "Failed to send file")
		Global.ReleaseJob(downloadID)
		return
	}
	defer f.Close()

	_, copyErr := io.Copy(w, f)

	if copyErr != nil {
		log.Printf("[%s] Stream error: %v", downloadID, copyErr)
		Global.SendProgressSimple(downloadID, "error", "Failed to send file")
		Global.ReleaseJob(downloadID)
	} else {
		Global.SendProgressSimple(downloadID, "complete", "Download complete!")
		Global.UnregisterDownload(downloadID)
		Global.ReleaseJob(downloadID)
	}

	if onCleanup != nil {
		onCleanup()
	}

	go func() {
		util.CleanupJobFiles(downloadID)
	}()
}

func ProbeForGif(filePath string) bool {
	cmd := exec.Command("ffprobe",
		"-v", "quiet",
		"-print_format", "json",
		"-show_streams", "-show_format",
		filePath,
	)
	out, err := cmd.Output()
	if err != nil {
		return false
	}

	var probe struct {
		Streams []struct {
			CodecType string `json:"codec_type"`
		} `json:"streams"`
		Format struct {
			Duration string `json:"duration"`
		} `json:"format"`
	}

	if err := json.Unmarshal(out, &probe); err != nil {
		return false
	}

	hasAudio := false
	for _, s := range probe.Streams {
		if s.CodecType == "audio" {
			hasAudio = true
			break
		}
	}

	duration := 999.0
	fmt.Sscanf(probe.Format.Duration, "%f", &duration)

	return !hasAudio && duration < 60
}

func GetMimeType(ext string, isAudio, isGif bool) string {
	if isGif {
		return "image/gif"
	}
	if isAudio {
		if mime, ok := config.AudioMIMEs[ext]; ok {
			return mime
		}
		return "audio/mpeg"
	}
	if mime, ok := config.ContainerMIMEs[ext]; ok {
		return mime
	}
	return "video/mp4"
}

func toASCII(s string) string {
	var b strings.Builder
	for _, r := range s {
		if r >= 0x20 && r <= 0x7E {
			b.WriteRune(r)
		} else {
			b.WriteRune('_')
		}
	}
	return b.String()
}
